### PR TITLE
Fix staging

### PR DESCRIPTION
**Pull request by Murmele**
- Hopefully closes also #106
- Add additional unittests
- 

The first should be selected, but it is the last. Don't know if this can be change.
But when clicking on the first deletion the last gets selected

image

### CODE REVIEW

File: src/ui/DiffView/HunkWidget.cpp [hunk 1]
```diff
@@ -869,16 +867,16 @@ void HunkWidget::setEditorLineInfos(QList<Line>& lines, Account::FileComments& c
...
     int marker = -1;
     int additions = 0, deletions = 0;
 	int additions_tot = 0, deletions_tot = 0;
     int stagedAdditions = 0, stagedDeletions = 0;
     bool staged = false;
 	int current_staged_index = -1;
 	int current_staged_line_idx = 0;
-    int diff_patch_old_new_file = 0; //-1;
-    int diff_staged_patch_old_new_file = 0; //-1;
+//	int diff_patch_old_new_file = 0;
```
Comment (exactly-one-kas) [inline src/ui/DiffView/HunkWidget.cpp:878]: Please remove these remnants

File: src/ui/DiffView/HunkWidget.cpp [hunk 2]
```diff
@@ -896,23 +894,26 @@ void HunkWidget::setEditorLineInfos(QList<Line>& lines, Account::FileComments& c
...
         // staged patch contains
 		if (!mPatch.isConflicted() && current_staged_index >= 0) {
 			auto staged_header_struct_next = mStaged.header_struct(current_staged_index + 1);
 			if (!first_staged_patch_match) {
-				if (mPatch.lineNumber(mIndex, lidx, git::Diff::OldFile) == mStaged.header_struct(current_staged_index)->old_start) {
+				if (mPatch.lineContent(mIndex, lidx) == mStaged.lineContent(current_staged_index, 0)) {
 					first_staged_patch_match = true;
-                    diff_patch_old_new_file = 0; //mPatch.lineNumber(mIndex, lidx, git::Diff::NewFile) - mPatch.lineNumber(mIndex, lidx, git::Diff::OldFile);
-                    diff_staged_patch_old_new_file = 0; //mStaged.lineNumber(current_staged_index, 0, git::Diff::NewFile) - mStaged.lineNumber(current_staged_index, 0, git::Diff::OldFile);
+//                    diff_patch_old_new_file = 0; //mPatch.lineNumber(mIndex, lidx, git::Diff::NewFile) - mPatch.lineNumber(mIndex, lidx, git::Diff::OldFile);
```
Comment (exactly-one-kas) [inline src/ui/DiffView/HunkWidget.cpp:915]: Also these

File: src/ui/DiffView/HunkWidget.cpp [hunk 3]
```diff
@@ -953,12 +954,12 @@ void HunkWidget::setEditorLineInfos(QList<Line>& lines, Account::FileComments& c
               deletions = 0;
             }
             // Check if staged
-
 			if (!mPatch.isConflicted() && mStaged.count() > 0 && current_staged_index >= 0 && current_staged_line_idx < mStaged.lineCount(current_staged_index)) {
 				auto line_origin = mStaged.lineOrigin(current_staged_index, current_staged_line_idx);
-				auto staged_file = mStaged.lineNumber(current_staged_index, current_staged_line_idx, git::Diff::NewFile) - diff_staged_patch_old_new_file;
-				auto patch_file = mPatch.lineNumber(mIndex, lidx, git::Diff::NewFile) - (additions_tot - stagedAdditions) + (deletions_tot - stagedDeletions) - diff_patch_old_new_file; // - offset_patch_old_new;
-                // TODO: try to avoid comparing strings!!!
+//				auto staged_file = mStaged.lineNumber(current_staged_index, current_staged_line_idx, git::Diff::NewFile) - diff_staged_patch_old_new_file;
```
Comment (exactly-one-kas) [inline src/ui/DiffView/HunkWidget.cpp:959]: Also these

File: test/EditorLineInfos.cpp [hunk 4]
```diff
@@ -1,7 +1,11 @@
 #include "Test.h"
 #include "dialogs/ExternalToolsDialog.h"
 
+#define private public
```
Comment (exactly-one-kas) [inline test/EditorLineInfos.cpp:4]: Maybe replace this with something akin to

```
#define public_for_tests public
```

and inside the headers use
```
#ifndef public_for_tests
    #define public_for_tests private
#endif
```

File: test/EditorLineInfos.cpp [hunk 5]
```diff
@@ -555,13 +600,154 @@ void TestEditorLineInfo::multipleHunks_misc1() {
...
+	stagedDiff = mRepo.diffTreeToIndex(commit.tree()); /* correct */
+	diff = mRepo.status(mRepo.index(), nullptr, false);
+	QVERIFY(stagedDiff.count() > 0);
+	QVERIFY(diff.count() > 0);
+	stagedPatch = stagedDiff.patch(0);
+
+	delete fw;
+	fw = new FileWidget(&diffView, diff, patch, stagedPatch, QModelIndex(), name, path, submodule);
+
+	// TODO: release fw
```
Comment (exactly-one-kas) [inline test/EditorLineInfos.cpp:639]: Isn't this TODO done in line 658?

**Review by exactly-one-kas — CHANGES_REQUESTED** [review]
Aside from a bit of cleanup that needs to be done and some preprocessor abuse this looks fine

File: src/ui/DiffView/HunkWidget.cpp (hunk 1 above)
Comment (Murmele) [reply src/ui/DiffView/HunkWidget.cpp:878]: done

File: src/ui/DiffView/HunkWidget.cpp (hunk 2 above)
Comment (Murmele) [reply src/ui/DiffView/HunkWidget.cpp:915]: done

File: src/ui/DiffView/HunkWidget.cpp (hunk 3 above)
Comment (Murmele) [reply src/ui/DiffView/HunkWidget.cpp:959]: done

File: test/EditorLineInfos.cpp (hunk 5 above)
Comment (Murmele) [reply test/EditorLineInfos.cpp:639]: thanks I forgott

**Comment by Murmele** [comment]
> Aside from a bit of cleanup that needs to be done and some preprocessor abuse this looks fine

Thanks for reviewing. For some reason the windows build failed

File: test/EditorLineInfos.cpp (hunk 4 above)
Comment (exactly-one-kas) [reply test/EditorLineInfos.cpp:4]: ~I really don't like these `#define` shenanigans. If you want, I can implement the changes I requested.~

```
EditorLineInfos.cpp.obj : error LNK2001: unresolved external symbol "public: void __cdecl HunkWidget::stageSelected(int,int,bool)" (?stageSelected@HunkWidget@@QEAAXHH_N@Z)
```
Looks like on Windows changing the visibility of methods / slots also affects name mangling and/or whether they get exported or not.
I think we should promote `stageSelected` to `public`

**Comment by exactly-one-kas** [comment]
> For some reason the windows build failed

Yep, I unresolved the `#define` review comment with additional information

File: test/EditorLineInfos.cpp (hunk 4 above)
Comment (Murmele) [reply test/EditorLineInfos.cpp:4]: I will have a look

Comment (Murmele) [reply test/EditorLineInfos.cpp]: done

**Comment by Murmele** [comment]
> > For some reason the windows build failed
> 
> Yep, I unresolved the `#define` review comment with additional information

sorry there I was too fast

**Review by exactly-one-kas — APPROVED** [review]

**Pull request merged**